pkg/agent/protocols/s3.go

```diff
@@ -0,0 +1,99 @@
+package protocols
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type S3Manager struct {
+	S3 *s3.S3
+}
+
+func (m *S3Manager) Download(modelDir string, storageUri string) error {
+	s3Uri := strings.TrimPrefix(storageUri, string(S3))
+	path := strings.SplitN(s3Uri, "/", 2)
+	bucket := path[0]
+	item := path[1]
+	s3ObjectLoactor := &S3ObjectDownloader{
+		ModelDir: modelDir,
+		Bucket: bucket,
+		Item: item,
+	}
+	objects, err := s3ObjectLoactor.GetAllObjects(m.S3)
+	if err != nil {
+		return fmt.Errorf("unable to get batch objects %v", err)
+	}
+	err = s3ObjectLoactor.Download(m.S3, objects)
+	if err != nil {
+		return fmt.Errorf("unable to get download objects %v", err)
+	}
+	return nil
+}
+
+var _ ProtocolManager = (*S3Manager)(nil)
+
+type S3ObjectDownloader struct {
+	ModelDir string
+	Bucket string
+	Item string
+}
+
+func (s *S3ObjectDownloader) GetAllObjects(s3Svc *s3.S3) ([]s3manager.BatchDownloadObject, error){
+	resp, err := s3Svc.ListObjects(&s3.ListObjectsInput{
+		Bucket:    aws.String(s.Bucket),
+		Prefix:    aws.String(s.Item),
+	})
+	if err != nil {
+		return nil, err
+	}
+	results := make([]s3manager.BatchDownloadObject, 0)
+
+	for _, object := range resp.Contents {
+		fileName := filepath.Join(s.ModelDir, *object.Key)
```

Comment:
Is key the basename of the path? What if the user passes us a storageuri with subfolders?

Reply:
this handles sub-folders, I can show a demo. But that is handled by the list objects command.